File: golang/08-jealous-function/kitchen_ticket.go
package kata

import "strings"

type TicketItem struct {
	Name      string
	Qty       int
	Modifiers []string
	Allergy   string
}
type Order struct {
	Items    []TicketItem
	Customer map[string]any
	Table    map[string]any
	Special  string
	Rush     bool
}

type KitchenTicket struct{}

func NewKitchenTicket() *KitchenTicket {
	return &KitchenTicket{}
}

func (kt *KitchenTicket) print_ticket(order Order) string {
	lines := []string{}
	lines = append(lines, "Table: "+kt.itoa(order.Table["number"].(int)))
	if order.Table["zone"] != nil {
		lines = append(lines, "Zone: "+order.Table["zone"].(string))
	} else {
		lines = append(lines, "Zone: main")
	}
	if order.Table["server"] != nil {
		lines = append(lines, "Server: "+order.Table["server"].(string))
	} else {
		lines = append(lines, "Server: unassigned")
	}
	lines = append(lines, "Customer: "+order.Customer["name"].(string))
	if order.Customer["vip"] == true {
		lines = append(lines, "VIP")
	}
	if order.Rush {
		lines = append(lines, "RUSH")
	}
	totalItems := 0
	for _, item := range order.Items {
		totalItems += item.Qty
		line := item.Name + " x" + kt.itoa(item.Qty)
		if len(item.Modifiers) > 0 {
			line += " [" + strings.Join(item.Modifiers, ", ") + "]"
		}
		if item.Allergy != "" {
			line += " ALLERGY:" + item.Allergy
		}
		lines = append(lines, line)
	}
	lines = append(lines, "Items: "+kt.itoa(totalItems))
	if len(order.Special) > 0 {
		lines = append(lines, "Special: "+order.Special)
	}
	lines = append(lines, "---")
	return strings.Join(lines, "\n")
}
func (kt *KitchenTicket) itoa(n int) string { return kt.fmtInt(n) }
func (kt *KitchenTicket) fmtInt(n int) string {
	if n == 0 {
		return "0"
	}
	digits := ""
	for n > 0 {
		digits = string(rune('0'+n%10)) + digits
		n /= 10
	}
	return digits
}

File: python/08-jealous-function/src/kitchen_ticket.py
class Order:
    def __init__(self, items, customer, table, special, rush=False):
        self.items = items
        self.customer = customer
        self.table = table
        self.special = special
        self.rush = rush


class KitchenTicket:
    def __init__(self):
        pass

    def print_ticket(self, order):
        lines = []
        lines.append("Table: " + str(order.table["number"]))
        lines.append("Zone: " + order.table.get("zone", "main"))
        lines.append("Server: " + order.table.get("server", "unassigned"))
        lines.append("Customer: " + order.customer["name"])
        if order.customer.get("vip"):
            lines.append("VIP")
        if order.rush:
            lines.append("RUSH")
        total_items = 0
        for item in order.items:
            total_items += item["qty"]
            line = item["name"] + " x" + str(item["qty"])
            if item.get("modifiers"):
                line += " [" + ", ".join(item["modifiers"]) + "]"
            if item.get("allergy"):
                line += " ALLERGY:" + item["allergy"]
            lines.append(line)
        lines.append("Items: " + str(total_items))
        if order.special and len(order.special) > 0:
            lines.append("Special: " + order.special)
        lines.append("---")
        return "\n".join(lines)
